### PR TITLE
Fixes #1071 - Web editor for repository files (Create new, edit, upload, rename, delete)

### LINKED ISSUE #1071: Web editor for repository files
so that folks can easily make edits to files in a git repository from their browser without having to clone a whole repo

### PR DESCRIPTION
**Pull request by richmahn**
NOTE: This also requires https://github.com/gogits/git-module/pull/12 for git-module for handling branches

NOTE: Discussion has already taken place on previous pull request #3092 (closed)

NOTE: I renamed font-awesome, semantics and simplemde just so we don't need the version numbers in path to them (they have version in their headers) so that in the templates/base/head.tmpl we don't need to change it, as that causes problems with custom/templates/base/head.tmpl if it has been copied there. I left the version number in the jquery library though.

Adds create new, edit, rename, uplad and delete feature for files

CodeMirror/SimpleMDE (same package) for editor
Moves workingPool out of the wiki.go file so it can be used for repos as well
Uses Dropzone to upload files
Feel free to code review and give constructive criticism. My first addition with Go Lang. Thanks.

### CODE REVIEW

**Comment by richmahn** [comment]
Again, this fails because it needs to check out my Pull Request for gogits/git-module

**Comment by jakimfett** [comment]
I haven't run into any problems while testing it. Looks great from a user perspective, can't wait to have this to my Gogs server!

**Comment by liebo** [comment]
readme md file is not rendered when view homepage of a repo

**Comment by richmahn** [comment]
@liebo I still haven't been able to reproduce this. Where did you see this? I'll make sure my test.door43.org:3000 copy is running the latest, but sometimes it isn't, and it has a bunch of other patches I have made for my organization's personal use of Gogs.

**Comment by Unknwon** [comment]
Tried the demo on test.door43.org:3000 and feels good. If this giant PR could break down to much smaller ones, would be easier to review and merge.

**Comment by richmahn** [comment]
@Unknwon Not sure how to break it down into different PRs as the code is all interconnected. I'd have to rework everything again and make new commits with new PRs.

File: models/action.go [hunk 1]
```diff
@@ -245,12 +245,12 @@ type PushCommits struct {
...
 	CompareUrl string
 
-	avatars map[string]string
+	Avatars map[string]string
 }
 
 func NewPushCommits() *PushCommits {
 	return &PushCommits{
-		avatars: make(map[string]string),
+		Avatars: make(map[string]string),
```
Comment (richmahn) [inline models/action.go]: This was uppercased to make it Public so other routers could make a PushCommits, such as line 66 of routers/repo/branch.go

File: models/repo.go [hunk 2]
```diff
@@ -2234,6 +2234,34 @@ func (repo *Repository) UpdateRepoFile(doer *User, oldBranchName, branchName, ol
 		return fmt.Errorf("Push: %v", err)
 	}
 
+	//p := &api.PushPayload{
```
Comment (richmahn) [inline models/repo.go]: This is removed in the "Code Cleanup" commit.

**Comment by Unknwon** [comment]
Because there are too many files just impossible for me to review.

**Comment by bkcsoft** [comment]
I can have a look at reviewing it, if @Unknwon is fine with that

**Comment by Unknwon** [comment]
Clean up the commits history and separate them properly (into smaller ones) could help review too, basically I need to make comments on GitHub.

**Comment by liebo** [comment]
@richmahn sorry for late response. I maintain a personal modification version of Gogs, and merged your patch. I have solved the problem of readme.md not rendered by modify modules/markdown/markdown.go

```
func IsMarkdownFile(name string) bool {
	name = strings.ToLower(name)
	extensions := []string{}
```

to
```
func IsMarkdownFile(name string) bool {
	name = strings.ToLower(name)
	extensions := []string{".md", ".markdown", ".mdown", ".mkd"}
```

now it works for me.

**Comment by richmahn** [comment]
@liebo  - this actually was to be done in the app.ini file, that is, can specify what files are to be rendered as markdown. I have fixed this in a new PR, as I am closing this one.

**Comment by richmahn** [comment]
Closed in favor of a split of PRs to PR #3286 and #3287

**Comment by richmahn** [comment]
@Unknwon I hope that just taking the SimpleMDE JS files out of it will help, as now it is just 35 files. So much of what I did, from branch creation to uploading files requires me to have it all as one PR, otherwise it wouldn't work by itself. Let me know if the new PRs, as listed above, allow you to get this into master. Thanks.

**Pull request closed without merging**